src/main/java/org/ossgang/commons/observable/mappers/buffer/ConcurrentCircularBuffer.java

```diff
@@ -0,0 +1,78 @@
+package org.ossgang.commons.observable.mappers.buffer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A circular buffer of a given size. The exact length is not fully guaranteed. Cleanup is done on each addition of a
+ * value. The length can be changed during runtime, however, the new length only applies as soon as a new element is
+ * added.
+ */
+public class ConcurrentCircularBuffer<T> {
+
+    private final AtomicLong nextIndex = new AtomicLong(0);
+    private final AtomicLong firstIndex = new AtomicLong(0);
+    private final AtomicInteger length = new AtomicInteger(1);
+
+    private final ConcurrentHashMap<Long, T> elements = new ConcurrentHashMap<>();
+
+    /**
+     * Adds a value to the buffer. If the length of the buffer is exceede, old elements are removed.
+     */
+    public void add(T value) {
+        long index = nextIndex.getAndIncrement();
+        elements.put(index, value);
+        cleanup();
+    }
+
+    /**
+     * Clears the buffer fully.
+     */
+    public void clear() {
+        cleanUpTo(nextIndex.get());
+    }
+
+    private void cleanup() {
+        long newFirstIndex = nextIndex.get() - length.get();
+        cleanUpTo(newFirstIndex);
+    }
+
+    private void cleanUpTo(long newFirstIndex) {
+        long oldFirstIndex = firstIndex.getAndSet(newFirstIndex);
+        for (long i = oldFirstIndex; i < newFirstIndex; i++) {
+            elements.remove(i);
+        }
+    }
+
+    /**
+     * Retrieves the current content of the buffer as a list.
+     */
+    public List<T> toList() {
+        List<T> list = new ArrayList<>();
+        for (long i = firstIndex.get(), next = nextIndex.get(); i < next; i++) {
+            T element = elements.get(i);
+            /*
+             * we have to check for null here, because a concurrent modification could have removed it in the meantime
+             */
+            if (element != null) {
```

Comment:
A rather conceptual question - what should happen if elements which we are accessing here are cleaned up while the loop runs?
I believe the current implementation does not guarantee that a List of exactly the length of the buffer is produced. It could be less, in case some early items were cleaned up before they were fetched. Probably this is as good as it can be, but this behavior should be documented.
An alternative could be to iterate while i < nextIndex.get() and list.size() < length?

Reply:
Without proper synchronization (not concurrent anymore) probably this is the best we can do.. this method could lead to misunderstandings since you could in principle get a buffer that is in between (possibly multiple) updates..

but I don't see another alternative while keeping it concurrent.. maybe having the cleanup and toList synchronized + michi's proposal will return a snapshot of the buffer that was the buffer itself at some point in time. Of course then the cleanup must happen on a background thread........ not sure if we want to go this deeply in the problem :smile:

Reply:
Yes it can be less in case some elements are removed while the loop runs OR if some elements were added while the loop runs.

I think the list.size() < length condition is not a good one. Indeed we want to be able to get the content of partially filled buffers.

for the i < nextIndex.get() ... I am a bit hesitating (I think I put on purpose in the init clause of the for() loop): If for some reason, the filling thread has a higher prio and is very fast, then this would mean a quasi endless loop for the toList() method ... With the current implementation we are at least sure we always get a snapshot ... But you are right ... In this case, we potentially would get always empty lists ;-) ... So not absolutely sure what to do here .. I keep it for the moment and comment .... If somebody feels motivated to improve ... please do.

Reply:
Ok Kajetan, if we properly document this we could leave it for the moment :)

Anyway I summarize the discussion here so we start an issue with this. The problem is that given the concurrent nature of the circular buffer we don't want to fully synchronize any operation. In this case the `#toList()` could have an undesired behaviour: shorter lists the buffer (possibly empty) or mixed content of the lists between buffers "states" in highly concurrent environments (list with "holes").
One possible solution is to synchronized the cleanup and the generation of the list plus making sure that the length of the list matches the buffer length. This would ensure that the returned list was a snapshot of the buffer at some point in time. This proposal complicates the implementation quite a lot since the cleanup will have to be async. In the end it could not be a bad idea in general so we can avoid to call the method on every insert but could, possibly, run periodically.

Reply:
See #23